src/shell.zig

```diff
@@ -0,0 +1,315 @@
+    errdefer shell.arena.allocator().free(res.stdout);
+
+    switch (res.term) {
+        .Exited => |code| if (code == 0) return error.NonZeroExitStatus,
+        else => return error.CommandFailed,
+    }
+
+    if (res.stderr.len > 0) return error.NonEmptyStderr;
+    return res.stdout;
+}
+
+/// Returns current git commit hash as an ASCII string.
+pub fn git_commit(shell: *Shell) ![40]u8 {
+    const stdout = try shell.exec_stdout("git rev-parse --verify HEAD", .{});
+    // +1 for trailing newline.
+    if (stdout.len != 40 + 1) return error.InvalidCommitFormat;
+    return stdout[0..40].*;
+}
+
+/// Returns current git tag.
+pub fn git_tag(shell: *Shell) ![]const u8 {
+    const stdout = try shell.exec_stdout("git describe --tags", .{});
+    return stdout;
+}
+
+const Argv = std.ArrayList([]const u8);
+
+/// Expands `cmd` into an array of command arguments, substituting values from `cmd_args`.
+///
+/// This avoids shell injection by construction as it doesn't concatenate strings.
+fn expand_argv(argv: *Argv, comptime cmd: []const u8, cmd_args: anytype) !void {
+    // Mostly copy-paste from std.fmt.format
+    comptime var pos: usize = 0;
+
+    comptime var args_used = 0;
+    inline while (pos < cmd.len) {
+        inline while (pos < cmd.len and cmd[pos] == ' ') {
+            pos += 1;
+        }
+
+        const pos_start = pos;
+        inline while (pos < cmd.len) : (pos += 1) {
+            switch (cmd[pos]) {
+                ' ', '{' => break,
+                else => {},
+            }
+        }
+
+        const pos_end = pos;
+        if (pos_start != pos_end) {
+            try argv.append(cmd[pos_start..pos_end]);
+        }
+
+        if (pos >= cmd.len) break;
+        if (cmd[pos] == ' ') continue;
+
+        comptime assert(cmd[pos] == '{');
+        if (pos > 0 and cmd[pos - 1] != ' ') @compileError("Surround {} with spaces");
+        pos += 1;
+
+        const pos_arg_start = pos;
+        inline while (pos < cmd.len and cmd[pos] != '}') : (pos += 1) {}
+        const pos_arg_end = pos;
+
+        if (pos >= cmd.len) @compileError("Missing closing }");
+
+        comptime assert(cmd[pos] == '}');
+        if (pos + 1 < cmd.len and cmd[pos + 1] != ' ') @compileError("Surround {} with spaces");
+        pos += 1;
+
+        const arg_name = comptime cmd[pos_arg_start..pos_arg_end];
+        const arg_or_slice = @field(cmd_args, arg_name);
+        args_used += 1;
+
+        if (@TypeOf(arg_or_slice) == []const u8) {
+            try argv.append(arg_or_slice);
+        } else if (@TypeOf(arg_or_slice) == []const []const u8) {
+            for (arg_or_slice) |arg_part| {
+                try argv.append(arg_part);
+            }
+        } else {
+            @compileError("Unsupported argument type");
+        }
+    }
+
+    const args_provided = std.meta.fields(@TypeOf(cmd_args)).len;
+    if (args_used != args_provided) @compileError("Unused argument");
+}
+
+test "format argv" {
```

Comment:
Sorry one more nit :D
```suggestion
test "shell: expand argv" {
```